azure/src/main/scala/quasar/blobstore/azure/AzureListService.scala

```diff
@@ -17,31 +17,43 @@
 package quasar.blobstore.azure
 
 import quasar.blobstore.azure.requests.ListBlobHierarchyArgs
-import quasar.blobstore.paths.PrefixPath
+import quasar.blobstore.paths.{BlobstorePath, PrefixPath}
 import quasar.blobstore.services.ListService
 
 import java.lang.Integer
 
 import scala.{None, Some}
 
 import cats.data.Kleisli
-import cats.effect.{Async, ContextShift}
-import com.microsoft.azure.storage.blob.{ContainerURL, ListBlobsOptions}
-import com.microsoft.rest.v2.Context
+import cats.effect.{ConcurrentEffect, ContextShift}
+import cats.syntax.functor._
+import cats.syntax.option._
+import com.azure.storage.blob.BlobContainerAsyncClient
+import com.azure.storage.blob.models.ListBlobsOptions
+import fs2.Stream
 
 object AzureListService {
-  def apply[F[_]: Async: ContextShift](
+
+  def apply[F[_]: ContextShift](
       toListBlobsOption: Kleisli[F, PrefixPath, ListBlobsOptions],
-      mkArgs: ListBlobsOptions => ListBlobHierarchyArgs)
+      mkArgs: ListBlobsOptions => ListBlobHierarchyArgs)(
+      implicit F: ConcurrentEffect[F])
       : ListService[F] =
     toListBlobsOption map
       mkArgs andThen
       requests.listRequestK andThen
-      converters.toBlobstorePathsK
-
-
-  def mk[F[_]: Async: ContextShift](containerURL: ContainerURL): ListService[F] =
+      converters.toBlobstorePathsK andThen
+      Kleisli { (s: Stream[F, BlobstorePath]) =>
+        // this is a workaround since this evaluates the stream twice
+        // but peek1 cannot be used since this results in an error in Azure
+        s.take(1).compile.last.map {
+          case None => None
+          case Some(_) => s.some
```

Comment:
I have tried with `peek1`, following Emrys's advice, but this gives an error in Azure which I think is a bug on their side and I haven't found anything to get around it. Let me know if you have further ideas for me to try. See for details: https://github.com/precog/async-blobstore/pull/96/commits/1cefe76376347706f7ff86661b36e6c4dd955831

I ended up reintroducing the `take(1)` approach but not from a generic function anymore and with documentation. I will also add a test case so that whenever something gets fixed or changes in this regard in azure sdk then we will get to know when we upgrade.

So what I think is happening with the `take(1)` approach is that the api call is done twice instead of once. This is of course undesired but not all that bad. In the other place where I introduced I added an "is existing"-check which also means an extra API call. I looked for an "is existing"-check for prefixes so that I could add it here too but haven't found it. I do see another solution and that is to look via the parent prefix but that's both more logic to write and less performant, so I chose `take(1)` instead.

All of this is assuming that `take(1)` indeed only means doing the call twice. LMK if I am missing something.

Reply:
@rintcius Looking at your `peek(1)` commit, the problem is due to calling `use`. Calling `use` on `Resource` requires that you fully "consume" whatever the `Resource` is managing, as the resource finalizer is guaranteed to run after the effect returned from the function passed to `use` terminates.

The solution here is to return the stream in `Resource[F, _]`, not in `F[_]` (this problem is precisely why we changed the effect that `Datasource` methods return to be `Resource[F, _]`)

Reply:
Ah! Thank you, I finally get it I think. The interface between async-blobstore and the datasource needs to change